### PR TITLE
[main] fix(telemetry): Filter non-plugin errors in Sentry

### PR DESCRIPTION
**Pull request by eguicf**
## Summary of Changes

This PR adds stack trace filtering to drop non-plugin errors in Sentry's automatic exception capture, preventing spike protection activation.

## Problem

After deploying v253.25910.0 on April 7, 2026, Sentry spike protection repeatedly [activated](https://confluent.slack.com/archives/C0A7SHY1U7K/p1775586958663009) and [deactivated](https://confluent.slack.com/archives/C0A7SHY1U7K/p1775592611660869) for the `intellij-plugin` project:

- **18:35 UTC** - Spike protection activated
- **20:10 UTC** - Deactivated
- **20:36 UTC** - Activated again
- **21:20 UTC** - Deactivated again

This cycling pattern indicated sustained, fluctuating error volume despite PR #388 implementing error filtering.

## Root Cause

**manual** error reports successfully filtered in #388 (when users click "Report to plugin vendor"), but missed a critical path:

**Automatic Capture (NOT Filtered)**
Uncaught exception anywhere in JVM
→ Sentry SDK's global UncaughtExceptionHandler
→ SentryClient.setBeforeSend() only added tags, didn't filter
→ ALL exceptions sent (plugin + platform)

These non-plugin errors continued flooding Sentry, triggering spike protection.

## Solution

Added stack trace filtering in `SentryClient.setBeforeSend()` to inspect and drop non-plugin errors **before** they reach Sentry:

### Changes:
1. **Created `SentryErrorFilter`** - Separated filtering logic to allow testing without triggering Sentry initialization
2. **Stack trace inspection** - Walks entire exception cause chain to detect plugin package (`io.confluent.intellijplugin.*`)
3. **Integrated filter in `setBeforeSend`** - Drops non-plugin events by returning `null`
4. Updated `PLUGIN_PACKAGE` to `CONFLUENT_PACKAGE ` to loose error filtering, **Centralized constant** - Moved to `TelemetryUtils` to keep filters in sync
5. **Added 11 test cases** - Covers plugin detection, package boundaries, wrapped exceptions, and edge cases

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

#### Tests

- [x] Added new
- [ ] Updated existing
- [ ] Deleted existing

### CODE REVIEW

File: test/io/confluent/intellijplugin/telemetry/SentryClientTest.kt [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                        "poll",
+                        "KafkaConsumer.java",
+                        123
+                    )
+                )
+            }
+
+            val event = SentryEvent(pluginException)
+            val isPluginError = SentryClient.isPluginRelatedError(event)
+
```
[the remark this answers is not in the document]
Comment (eguicf) [reply test/io/confluent/intellijplugin/telemetry/SentryClientTest.kt:43]: 👍 updated

File: src/io/confluent/intellijplugin/telemetry/SentryClient.kt [hunk 2]
```diff
@@ -18,8 +20,14 @@ object SentryClient {
...
                 options.serverName = TelemetryUtils.getAnonymisedHostname()
                 options.setBeforeSend { event, _ ->
-                    addDefaultTags(event)
-                    event
+                    // Filter out non-plugin errors before sending to Sentry
+                    if (isPluginRelatedError(event)) {
+                        addDefaultTags(event)
+                        event
+                    } else {
+                        logger.info("Dropping non-plugin error from Sentry: ${event.throwable?.javaClass?.name}")
```
[the remark this answers is not in the document]
Comment (eguicf) [reply src/io/confluent/intellijplugin/telemetry/SentryClient.kt]: updated

File: src/io/confluent/intellijplugin/telemetry/SentryClient.kt [hunk 3]
```diff
@@ -5,9 +5,11 @@ import com.intellij.openapi.application.PermanentInstallationID
 import com.intellij.openapi.diagnostic.Logger
 import com.intellij.openapi.util.SystemInfo
 import io.sentry.Sentry
+import io.sentry.SentryEvent
 
 object SentryClient {
     private val logger = Logger.getInstance(SentryClient::class.java)
+    private const val PLUGIN_PACKAGE = "io.confluent.intellijplugin"
```
[the remark this answers is not in the document]
Comment (eguicf) [reply src/io/confluent/intellijplugin/telemetry/SentryClient.kt:11]: Centralized in `TelemetryUtils`.

File: src/io/confluent/intellijplugin/telemetry/SentryClient.kt [hunk 4]
```diff
@@ -48,4 +56,19 @@ object SentryClient {
...
+     *
+     * @param event The Sentry event to check
+     * @return true if the error is plugin-related, false otherwise
+     */
+    internal fun isPluginRelatedError(event: SentryEvent): Boolean {
+        val throwable = event.throwable ?: return false
+
+        // Check if any stack trace element contains the plugin package
+        return throwable.stackTrace.any { frame ->
+            frame.className.startsWith(PLUGIN_PACKAGE)
```
[the remark this answers is not in the document]
Comment (eguicf) [reply src/io/confluent/intellijplugin/telemetry/SentryClient.kt]: updated

File: src/io/confluent/intellijplugin/telemetry/SentryClient.kt [hunk 5]
```diff
@@ -48,4 +56,19 @@ object SentryClient {
...
+        }
```
[the remark this answers is not in the document]
Comment (eguicf) [reply src/io/confluent/intellijplugin/telemetry/SentryClient.kt]: good point, updated

File: test/io/confluent/intellijplugin/telemetry/SentryClientTest.kt [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                    )
+                )
+            }
+
+            val event = SentryEvent(similarPackageException)
+            val isPluginError = SentryErrorFilter.isPluginRelatedError(event)
+
+            assertFalse(
+                isPluginError,
+                "Should reject packages that start with io.confluent but aren't the plugin"
```
Comment (noeldevelops) [inline test/io/confluent/intellijplugin/telemetry/SentryClientTest.kt]: I disagree with this statement. Maybe we need to loosen the requirement here - it seems like if the error contains `confluent` at all we should send it to Sentry?

**Review by noeldevelops — COMMENTED** [review]
the logic looks solid to me and the tests give me confidence. 

I left a comment about loosening the filter requirements to just check for `confluent` but I'm not sure about it. What do y'all think?

File: test/io/confluent/intellijplugin/telemetry/SentryClientTest.kt (hunk 6 above)
Comment (eguicf) [reply test/io/confluent/intellijplugin/telemetry/SentryClientTest.kt]: That is a good point! Let's track all errors from `io.confluent`. I currently did not see many errors from `io.confluent` that is not the plugin related. If they are too noisy in the next month, we can always change back.

File: src/io/confluent/intellijplugin/telemetry/SentryErrorFilter.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            }
+
+            // Check if any stack frame belongs to Confluent package
+            if (current.stackTrace.any { frame ->
+                frame.className.startsWith(TelemetryUtils.CONFLUENT_PACKAGE_PREFIX)
+            }) {
+                return true
+            }
+
+            // Check if exception message mentions Confluent package
```
Comment (eguicf) [inline src/io/confluent/intellijplugin/telemetry/SentryErrorFilter.kt:43]: Added this part becuase some errors does not include `io.confluent(.intellijplugin)` in the stack trace, instead, only in the exception message. ([example](https://confluent.sentry.io/issues/7103510176/?project=4510160131588096&query=&referrer=issue-stream))

Comment (noeldevelops) [reply src/io/confluent/intellijplugin/telemetry/SentryErrorFilter.kt:43]: ty!

**Review by noeldevelops — APPROVED** [review]
looks good; let's try it out ☘️
